src/main/java/hudson/plugins/ec2/SlaveTemplate.java

```diff
@@ -1164,44 +1167,68 @@ private List<EC2AbstractSlave> provisionOndemand(Image image, int number, EnumSe
         wakeOrphansOrStoppedUp(ec2, orphansOrStopped);
 
         if (orphansOrStopped.size() == number) {
-            return toSlaves(orphansOrStopped);
+            return toSlaves(InstanceInfo.fromInstances(orphansOrStopped, getParent()));
         }
 
         riRequest.setMaxCount(number - orphansOrStopped.size());
 
-        List<Instance> newInstances;
+        List<InstanceInfo> newInstances;
         if (spotWithoutBidPrice) {
             InstanceMarketOptionsRequest instanceMarketOptionsRequest = new InstanceMarketOptionsRequest().withMarketType(MarketType.Spot);
             if (getSpotBlockReservationDuration() != 0) {
                 SpotMarketOptions spotOptions = new SpotMarketOptions().withBlockDurationMinutes(getSpotBlockReservationDuration() * 60);
                 instanceMarketOptionsRequest.setSpotOptions(spotOptions);
             }
             riRequest.setInstanceMarketOptions(instanceMarketOptionsRequest);
+
             try {
-                newInstances = ec2.runInstances(riRequest).getReservation().getInstances();
+                newInstances = createNewInstances(ec2, riRequest);
             } catch (AmazonEC2Exception e) {
                 if (fallbackSpotToOndemand && e.getErrorCode().equals("InsufficientInstanceCapacity")) {
                     logProvisionInfo("There is no spot capacity available matching your request, falling back to on-demand instance.");
                     riRequest.setInstanceMarketOptions(new InstanceMarketOptionsRequest());
-                    newInstances = ec2.runInstances(riRequest).getReservation().getInstances();
+                    newInstances = createNewInstances(ec2, riRequest);
                 } else {
                     throw e;
                 }
             }
         } else {
-            newInstances = ec2.runInstances(riRequest).getReservation().getInstances();
+            newInstances = createNewInstances(ec2, riRequest);
         }
         // Have to create a new instance
 
         if (newInstances.isEmpty()) {
             logProvisionInfo("No new instances were created");
         }
 
-        newInstances.addAll(orphansOrStopped);
+        newInstances.addAll(InstanceInfo.fromInstances(orphansOrStopped, getParent()));
 
         return toSlaves(newInstances);
     }
 
+    List<InstanceInfo> createNewInstances(AmazonEC2 ec2, RunInstancesRequest riRequest) {
+        if (isUsingDynamicSshKeys()) {
+            // no static ssh key defined, so submit riRequest.maxCount requests
+            // each with a unique keypair (instead of a single request for multiple identical instances)
+            LOGGER.fine(() -> "no static ssh credential configured, will use dynamic ssh keys instead");
+            List<InstanceInfo> instances = new ArrayList<>();
+            int maxRequested = riRequest.getMaxCount();
+            riRequest.setMaxCount(1);
```

Comment:
Looks correct but surprising—you do not normally expect some object like this to be mutated by a subroutine. While I guess `RunInstancesRequest` is just some mutable bean that gets serialized anew for every `runInstances` call, it would be clearer if the caller of this method were to pass in the necessary parameters for a `RunInstancesRequest` (or a `Supplier<RunInstancesRequest>`) so that a fresh request object would be constructed on each API call.

Reply:
I can take a look at this later today, although i don't promise it will be in this PR

Reply:
not implemented